### PR TITLE
LCI – ReplaceExplore cards and support

### PR DESCRIPTION
**Pull request by Northmoc**

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/topography_tracker.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Name:Topography Tracker
+ManaCost:2 G
+Types:Creature Merfolk Scout
+PT:2/2
+T:Mode$ ChangesZone | Origin$ Any | Destination$ Battlefield | ValidCard$ Card.Self | Execute$ TrigToken | TriggerDescription$ When CARDNAME enters the battlefield, create a Map token. (It's an artifact with "{1}, {T}, Sacrifice this artifact: Target creature you control explores. Activate only as a sorcery.")
+SVar:TrigToken:DB$ Token | TokenScript$ c_a_map_sac_explore
+R:Event$ Explore | ValidExplorer$ Creature.YouCtrl | ReplaceWith$ Explore1 | Description$ If a creature you control would explore, instead it explores, then it explores again.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/topography_tracker.txt:7]: Zones missing?

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/topography_tracker.txt:7]: I always forget those when I write R: lines from scratch

File: forge-gui/res/cardsfolder/upcoming/topography_tracker.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+SVar:Explore1:DB$ Explore | Defined$ ReplacedCard | SubAbility$ Explore2
```
Comment (Hanmac) [inline forge-gui/res/cardsfolder/upcoming/topography_tracker.txt:8]: You could still write it as Explore Num$ 2

File: forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java [hunk 3]
```diff
@@ -45,54 +47,60 @@ protected String getStackDescription(SpellAbility sa) {
...
+                    Card movedCard = null;
+                    game.getAction().reveal(top, pl, false,
+                            Localizer.getInstance().getMessage("lblRevealedForExplore") + " - ");
+                    final Card r = top.getFirst();
+                    final Zone originZone = game.getZoneOf(r);
+                    if (r.isLand()) {
+                        movedCard = game.getAction().moveTo(ZoneType.Hand, r, sa, moveParams);
+                        revealedLand = true;
+                    } else {
+                        if (pl.getController().confirmAction(sa, null,
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java:80]: @Agetian this is the one 😉

File: forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java [hunk 4]
```diff
@@ -45,54 +47,60 @@ protected String getStackDescription(SpellAbility sa) {
...
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+        int amount = AbilityUtils.calculateAmount(host, sa.getParamOrDefault("Num", "1"), sa);
 
         GameEntityCounterTable table = new GameEntityCounterTable();
         final CardZoneTable triggerList = new CardZoneTable();
         Map<AbilityKey, Object> moveParams = AbilityKey.newMap();
         moveParams.put(AbilityKey.LastStateBattlefield, sa.getLastStateBattlefield());
         moveParams.put(AbilityKey.LastStateGraveyard, sa.getLastStateGraveyard());
         for (final Card c : getTargetCards(sa)) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java:59]: The changeZone and Counter Trigger needs to be inside the For Amount Loop

so "Once" Replacement/Trigger would trigger multiple times

File: forge-game/src/main/java/forge/game/player/PlayerController.java [hunk 5]
```diff
@@ -131,7 +131,10 @@ public void endTempShowCards() { }
 
     public abstract <T extends GameEntity> List<T> chooseEntitiesForEffect(FCollectionView<T> optionList, int min, int max, DelayedReveal delayedReveal, SpellAbility sa, String title, Player relatedPlayer, Map<String, Object> params);
 
-    public abstract boolean confirmAction(SpellAbility sa, PlayerActionConfirmMode mode, String message, Map<String, Object> params);
+    public final boolean confirmAction(SpellAbility sa, PlayerActionConfirmMode mode, String message, Map<String, Object> params) {
+        return confirmAction(sa, mode, message, params, null);
+    }
+    public abstract boolean confirmAction(SpellAbility sa, PlayerActionConfirmMode mode, String message, Map<String, Object> params, Card cardToShow);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/player/PlayerController.java:137]: i would prefer if the Map is the Last param, but i think that's okay for now

**Comment by Agetian** [comment]
I have a working version of the AI for this branch, but I can't push it upstream, I don't have permissions to push directly into this branch. I may need to wait for this to go into the upstream Forge, I guess?.. Or is it possible to give me the permission to commit and push here?

**Review by tool4ever — APPROVED** [review]

**Review by Agetian — APPROVED** [review]

File: forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java (hunk 3 above)
Comment (Agetian) [reply forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java:84]: Done :)

File: forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java (hunk 4 above)
Comment (Agetian) [reply forge-game/src/main/java/forge/game/ability/effects/ExploreEffect.java]: Is this one covered, or does it need more work?

**Pull request merged**